main.go

```diff
@@ -139,19 +139,34 @@ func start_app(config Config) {
 	if config.Trusted_Root_Ca_File != "" {
 		content, err := ioutil.ReadFile(config.Trusted_Root_Ca_File)
 		if err != nil {
-			log.Fatalf("Failed to read file Trusted Root CA %s", config.Trusted_Root_Ca_File)
+			log.Fatalf("Failed to read file Trusted Root CA %s, %v", config.Trusted_Root_Ca_File, err)
 		}
 		ok := certp.AppendCertsFromPEM([]byte(content))
 		if !ok {
 			log.Fatalf("Failed to parse a trusted cert from file %s, pem format expected", config.Trusted_Root_Ca_File)
 		}
 	}
 
-	mTlsConfig := &tls.Config{}
-	mTlsConfig.PreferServerCipherSuites = true
-	mTlsConfig.MinVersion = tls.VersionTLS10
-	mTlsConfig.MaxVersion = tls.VersionTLS12
-	mTlsConfig.RootCAs = certp
+	mTlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12, // minimum TLS 1.2
+		// P curve order does not matter, as breaking one means all others can be brute-forced as well:
+		// Golang developers prefer:
+		CurvePreferences:         []tls.CurveID{tls.X25519, tls.CurveP256, tls.CurveP384, tls.CurveP521},
+		PreferServerCipherSuites: true, // Server chooses ciphersuite, order matters below:
+		CipherSuites: []uint16{
```

Comment:
@Cryptophobia I was doing some background reading: https://github.com/ssllabs/research/wiki/SSL-and-TLS-Deployment-Best-Practices

In the Go community, is there more concise guidance on what we should be using? Is there a published list somewhere?

Thanks

Reply:
Hi @nabadger ,

This is a question of curve preference and TLS protocol support between servers and clients when they negotiate with each other. The Golang team does not set those standards, it is really supposed to based on the amount of security of the environments in which dex-k8s-authenticator will be deployed. These choices are usually just a tradeoff between performance and security. The curves and ciphers have different performance on different browsers and hardware. Golang devs are consumers of openssl ciphersuites and the cryptography community, so they are not opinionated about it very much.

The above comment `Golang developers prefer:` is based on my reading and understanding of this issue comment, which is a great read by the way: https://github.com/syncthing/syncthing/issues/5940

Since most of the time dex-k8s-authenticator will be used on desktop browsers (and not mobile devices - particularly ARM cpu architecture) to pull k8s configs, it makes sense for example that we should not prefer `CHACHA20_POLY1305` over `AES256-GCM`. Good read about this: https://blog.cloudflare.com/do-the-chacha-better-mobile-performance-with-cryptography/

Anyways, it is really just a matter of curve preference. And I think for a tool like this. We should always prefer maximum security at the expense of performance, unless there are special needs to optimize performance over security.

Cheers! 🍻

Reply:
Btw, we are already running in production environments with these settings so it did not break anything for us, it should be a pretty safe change for everyone. It is one step forward in preferring and moving towards TLS1.3 so it is good for everyone in general.

Reply:
Thanks for the response, some great reading for me! :)